Bug fix

cdawei opened the pull request:
# Pull Request Details

## What issue does this PR address

- Merge bug fix in pull request #247 into main branch

- Link to associated issue: #246 

## Checklist

Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [ ] Screenshots included in linked issue
- [x] Changes adhere to the team style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [ ] No lint check errors related to your changes (`make prep` or `flutter analyze lib`)
- [ ] Pre-exisiting lint errors noted: [HERE]
- [x] Tested on at least one device
  - [ ] Android Phone
  - [ ] Android Emulator
  - [ ] Chrome on Android
  - [ ] Chrome
  - [ ] iOS
  - [x] Linux
  - [ ] MacOS
  - [x] Windows
- [x] Added 2 reviewers (or 1 for private repositories then they add another)

## Finalising

Once PR discussion is complete and 2 reviewers have approved:

- [x] Merge dev into the branch
- [x] Resolve any conflicts
- [x] Add one line summary into CHANGELOG.md
- [x] Bump appropriate version number in pubspec.yaml
- [x] Push to git repository and review
- [x] Merge PR into dev

anushkavidanage commented:
Tested and thanks @cdawei

cdawei commented:
> Tested and thanks @cdawei

Thanks @anushkavidanage. Are you able to upload this version to pub.dev (after merging this pull request)?

anushkavidanage commented:
Hi @cdawei, Graham has access to update the version in pub.dev.

cdawei commented:
> Hi @cdawei, Graham has access to update the version in pub.dev.

Thanks @anushkavidanage. I will let Graham know once @srodriguez142857 approves the pull request.

srodriguez142857 commented:
I apologise for the delay in testing this PR.
I thought I did before, but I realise it's a different one... 🤓 (The one I reviewed before was for PR #247).

---

I did some testing. Interestingly, I found one error related to the `key-value.ttl` file.
For some reason, it gave the following errors the first time when trying to save the file.

```
flutter: server started 4400
flutter: AuthDataManager => saveAuthData() done
flutter: Security key saved
flutter: Invalid argument(s): Invalid or corrupted pad block
flutter: Resource "exampleApp/data/key-value.ttl" does not exist.
flutter: Invalid argument(s): Invalid or corrupted pad block
flutter: Resource "exampleApp/data/key-value.ttl" does not exist.
[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: Invalid argument(s): Invalid or corrupted pad block
#0      PKCS7Padding.padCount (package:pointycastle/paddings/pkcs7.dart:42:7)
#1      PaddedBlockCipherImpl.doFinal (package:pointycastle/padded_block_cipher/padded_block_cipher_impl.dart:112:30)
#2      PaddedBlockCipherImpl.process (package:pointycastle/padded_block_cipher/padded_block_cipher_impl.dart:74:25)
#3      AES.decrypt (package:encrypt/src/algorithms/aes.dart:68:22)
#4      Encrypter.decryptBytes (package:encrypt/src/encrypter.dart:39:10)
#5      Encrypter.decrypt (package:encrypt/src/encrypter.dart:50:7)
#6      decryptData (package:solidpod/src/solid/utils/misc.dart:96:37)
#7      KeyManager.getIndividualKey (package:solidpod/src/solid/utils/key_helper.dart:443:9)
<asynchronous suspension>
#8      writePod (package:solidpod/src/solid/write_pod.dart:148:7)
<asynchronous suspension>
#9      _KeyValueEditState._saveToPod (package:demopod/features/edit_keyvalue.dart:171:28)
<asynchronous suspension>
#10     _KeyValueEditState.build.<anonymous closure> (package:demopod/features/edit_keyvalue.dart:224:39)
<asynchronous suspension>

[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: type 'Null' is not a subtype of type 'String' in type cast
#0      _KeyValueEditState._saveEditedRows (package:demopod/features/edit_keyvalue.dart:114:41)
#1      _KeyValueEditState._saveToPod (package:demopod/features/edit_keyvalue.dart:149:5)
#2      _KeyValueEditState.build.<anonymous closure> (package:demopod/features/edit_keyvalue.dart:224:45)
#3      _InkResponseState.handleTap (package:flutter/src/material/ink_well.dart:1176:21)
#4      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:351:24)
#5      TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:656:11)
#6      BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:313:5)
#7      BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:246:7)
#8      PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:703:9)
#9      PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:98:12)
#10     PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:143:9)
#11     _LinkedHashMapMixin.forEach (dart:_compact_hash:726:13)
#12     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:141:18)
#13     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:127:7)
#14     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:502:19)
#15     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:482:22)
#16     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:457:11)
#17     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:427:7)
#18     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:390:5)
#19     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:337:7)
#20     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:306:9)
#21     _invoke1 (dart:ui/hooks.dart:328:13)
#22     PlatformDispatcher._dispatchPointerDataPacket (dart:ui/platform_dispatcher.dart:450:7)
#23     _dispatchPointerDataPacket (dart:ui/hooks.dart:262:31)

flutter: Invalid argument(s): Invalid or corrupted pad block
flutter: Resource "exampleApp/data/key-value.ttl" does not exist.
flutter: ACL file for "https://pods.solidcommunity.au/sergio-test2-POD/exampleApp/data/key-value.ttl" does not exist.
flutter: Deleted IndividualKeyRecord {
flutter:     filePath: exampleApp/data/key-value.ttl,
flutter:     encIndKey: ieqgdXs40Kh/JjgKYix6CohKuLD0g4IcfR3KHyyaDvpOhLvCqyyCyhomhAt2O1+W,
flutter:     iv: 9fMW79mbeS2FOUTLjGy4Kg==
flutter: }
flutter: Resource "exampleApp/data/key-value.ttl" does not exist.
```

I proceeded to delete the file and try again.  I didn't get the error afterwards.
Is this expected?

cdawei commented:
Hi @srodriguez142857, thanks for reporting this issue.

> flutter: Invalid argument(s): Invalid or corrupted pad block
> flutter: Resource "exampleApp/data/key-value.ttl" does not exist.
> flutter: ACL file for "https://pods.solidcommunity.au/sergio-test2-POD/exampleApp/data/key-value.ttl" does not exist.
> flutter: Deleted IndividualKeyRecord {
> flutter:     filePath: exampleApp/data/key-value.ttl,
> flutter:     encIndKey: ieqgdXs40Kh/JjgKYix6CohKuLD0g4IcfR3KHyyaDvpOhLvCqyyCyhomhAt2O1+W,
> flutter:     iv: 9fMW79mbeS2FOUTLjGy4Kg==
> flutter: }

I can see two problems from this message:
1. The file `key-value.ttl` does not exist, but the app can find its corresponding encryption key (from `exampleApp/encryption/ind-keys.ttl`). It could be that the `key-value.ttl` file was somehow deleted.
2. The encryption key for the (missing) `key-value.ttl` was corrupted.

So the data stored in the POD was inconsistent when you started to test the app, however, after you deleted the file (likely by clicking the button `Delete Pod Data File`), the (corrupted) encryption key for the `key-value.ttl` file was also deleted. When you tried again, a new encryption key was generated for the `key-value.ttl` file, as a result, the data store in the POD becomes consistent.

> I proceeded to delete the file and try again. I didn't get the error afterwards.
> Is this expected?

This seems understandable and follows the logic in the code, I would say it's an expected behaviour.

srodriguez142857 commented:
Thanks for the detailed response, @cdawei.
Thus, I don't have any further observations and the app is working as expected with my test.
All good. 🙂

srodriguez142857 reviewed (approved):
The `demopod` app is working as expected.
